### PR TITLE
initial commit of separation of business-facing datalake methods

### PR DESCRIPTION
**Pull request by nlawre21**
**New Features**
- Scratch partner support
- Python 3 Compatible
- Operation protection (i.e. dropping of production or non-owned scratch tables)
- Active-Directory username and execution environment showing on Presto Queries 
- Simplification of classes and repository structure (14k LOC to 1k) 

**Concepts and Design Decisions**
- 4 user-facing classes (Qubole Connection, Presto Connection, SQLConstructor, Table)
- Table class delegates to Qubole Connection and SQLConstructor. (if a good Hive ORM existed, it could replace this delegation behaviour)
- SQLConstructor just consolidates sql string literals, provides location to check queries before their execution.
- Important decision made is that all Qubole query results are retrieved from S3, as opposed to only large > 20mb results. This decision was made because 

1. it removed need to determine the size of the result
2. removed parameterization of delimiters (delimiters changed depending on if results were retrieved directly from Qubole vs S3)
3. Removed need to conditionally convert datatypes. Instead all data regardless of result size undergoes same conversion.

- By far biggest obstacle is reading data in from Qubole HiveCommand into a Pandas Dataframe and honoring the datatypes. This is an obstacle because the decision to get all data from S3 means all data is returned as a string. We leverage the datatype schema returned from HiveCommand to map to Pandas datatypes.

### CODE REVIEW

File: .buildkite/test [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+pip3 install -r requirements.txt --extra-index-url=http://pypi.ds.avant.com/ --trusted-host pypi.ds.avant.com
+
+echo "--- [Installing Pythons]"
+pyenv install 2.7.14
+pyenv install 3.6.3
+pyenv local 2.7.14 3.6.3
+
+echo "+++ [Running tests]"
+python2 -m tests.test_datalake
+python3 -m tests.test_datalake
```
Comment (samdjohnson) [inline .buildkite/test]: you'll have to explain this to me.

File: tests/test_datalake.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import unittest
+
+from data_lake_tools import datalake
+
+
+class TestConnect(unittest.TestCase):
+
+    def test_right_class_returned(self):
+        q_expected = datalake.QuboleDB()
+        q_actual = datalake.connect('qubole')
```
Comment (samdjohnson) [inline tests/test_datalake.py]: you should patch the trellis call so we aren't calling to an external service

File: .buildkite/test [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+echo "--- [Installing Libraries]"
+pip2 install --upgrade pip
+pip2 install -r requirements.txt --extra-index-url=http://pypi.ds.avant.com/ --trusted-host pypi.ds.avant.com
+
+pip3 install --upgrade pip
+pip3 install -r requirements.txt --extra-index-url=http://pypi.ds.avant.com/ --trusted-host pypi.ds.avant.com
+
+echo "--- [Installing Pythons]"
+pyenv install 2.7.14
+pyenv install 3.6.3
```
Comment (samdjohnson) [inline .buildkite/test]: you may be able to reference the existing python installation

File: data_lake_tools/__init__.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+__title__ = "data-lake-tools"
+__version__ = "1.0"
+__author__ = "Avant Data Services Team"
+__license__ = "MIT"
```
Comment (samdjohnson) [inline data_lake_tools/__init__.py]: please remove the license from this. it's proprietary code.

File: .buildkite/test (hunk 1 above)
Comment (nlawre21) [reply .buildkite/test]: specifically the py2 and py3 parts? or the file as a whole

File: data_lake_tools/__init__.py (hunk 4 above)
Comment (nlawre21) [reply data_lake_tools/__init__.py]: done

File: data_lake_tools/datalake.py [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+
+        return presto.connect(
+                host,
+                port=int(port),
+                protocol=protocol,
+                verify_ssl=verify_ssl,
+                username=self._platform + '-' + self._user
+        )
+
+    def execute_df(self, query, params=None):
```
[the remark this answers is not in the document]
Comment (nlawre21) [reply data_lake_tools/datalake.py]: * add execute metho back

File: data_lake_tools/datalake.py [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+
+            if "trying to fetch results on an operation with no results" in err:
+                results = pd.DataFrame()
+            else:
+                raise
+
+        return results
+
+
+class QuboleDB(DB):
```
Comment (nlawre21) [inline data_lake_tools/datalake.py]: rename to HiveDB

Comment (nlawre21) [reply data_lake_tools/datalake.py]: abstract out to be agnostic of provider

File: data_lake_tools/datalake.py [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+
+        :param query:             str  = Hive query to be executed, and whose dates are to be returned in a pd.Dataframe
+        :param convert_datatypes: bool = Indicate if columns should be converted using Qubole query result schema,
+                                         consideration should be given to additional execution time. If false, all
+                                         columns are objects.
+
+        :return:
+        """
+
+        assert query.strip().lower().startswith(
```
Comment (nlawre21) [inline data_lake_tools/datalake.py]: replace startswith with find

File: data_lake_tools/datalake.py [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+                ('with', 'select', 'describe', 'show')), "Please enter a command expecting results"
+
+        if params and isinstance(params, dict):
+            query = self._inject_params(query, params)
```
Comment (nlawre21) [inline data_lake_tools/datalake.py]: possibly look for hive parameter before executing

File: data_lake_tools/datalake.py [hunk 9]
```diff
@@ -0,0 +1,518 @@
...
+    def get_table_dtypes(self):
+        """
+        Return column -> datatype dictionary of table datatypes. OrderedDict is used to maintain column order.
+        Existence of data is understood to indicate existence of table.
+        """
+        describe_table = self._sql.gen_describe_table()
+        try:
+            dtype_df = self._qubole.execute_df(describe_table)
+            return OrderedDict(zip(dtype_df.col_name, dtype_df.data_type))
+        except (S3ResponseError, DatalakeError):
```
Comment (nlawre21) [inline data_lake_tools/datalake.py]: s3responseerror is too broad, filter by code/message

File: data_lake_tools/datalake.py [hunk 10]
```diff
@@ -0,0 +1,518 @@
...
+
+        :param overwrite   : bool = instruct to empty table directory before uploading data.
+        :param cleanup     : bool = remove locally saved file after upload.
+        :return:
+
+        """
+
+        if not self.table_dtypes:
+            assert data_types, "If the table doesn't exist you must provide datatypes!"
+            self._create_table(file_format, delim, data_types)
```
Comment (nlawre21) [inline data_lake_tools/datalake.py]: handle boolean datatypes, warn about nulls/empty strings

File: data_lake_tools/datalake.py (hunk 7 above)
Comment (nlawre21) [reply data_lake_tools/datalake.py]: done

File: data_lake_tools/datalake.py (hunk 6 above)
Comment (nlawre21) [reply data_lake_tools/datalake.py]: done

File: data_lake_tools/datalake.py (hunk 10 above)
Comment (nlawre21) [reply data_lake_tools/datalake.py]: added warn

**Pull request merged**